src/envelope.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use crate::errors::{InternalPakeError, PakeError};
+use generic_array::{
+    typenum::{Unsigned, U32},
+    GenericArray,
+};
+use hkdf::Hkdf;
+use hmac::{Hmac, Mac, NewMac};
+use rand_core::{CryptoRng, RngCore};
+use sha2::{Digest, Sha256};
+
+// Constant string used as salt for HKDF computation
+const STR_ENVU: &[u8] = b"EnvU";
+
+/// The length of the "export key" output by the client registration
+/// and login finish steps
+pub(crate) type ExportKeySize = <Sha256 as Digest>::OutputSize;
+
+/// This struct is a straightforward instantiation of the trait separating the
+/// three components in Vecs
+pub(crate) struct Envelope {
+    nonce: Vec<u8>,
+    ciphertext: Vec<u8>,
+    hmac: Vec<u8>,
+}
+
+type NonceLen = U32;
+
+impl Envelope {
+    /// The additional number of bytes added to the plaintext
+    pub(crate) fn additional_size() -> usize {
+        Self::nonce_size() + <Sha256 as Digest>::OutputSize::to_usize()
+    }
+
+    fn hmac_key_size() -> usize {
+        <Sha256 as Digest>::OutputSize::to_usize()
+    }
+
+    fn hmac_size() -> usize {
+        <Sha256 as Digest>::OutputSize::to_usize()
+    }
+
+    fn nonce_size() -> usize {
+        NonceLen::to_usize()
+    }
+
+    fn export_key_size() -> usize {
+        ExportKeySize::to_usize()
+    }
+
+    pub(crate) fn new(
+        nonce: Vec<u8>,
+        ciphertext: Vec<u8>,
+        hmac: &GenericArray<u8, <Sha256 as Digest>::OutputSize>,
+    ) -> Self {
+        Self {
+            nonce,
+            ciphertext,
+            hmac: hmac.to_vec(),
+        }
+    }
+
+    /// The format of the output is:
+    /// nonce            | ciphertext       | hmac
+    /// nonce_size bytes  | variable length  | hmac_size bytes
+    pub(crate) fn from_bytes(bytes: &[u8]) -> Result<Self, InternalPakeError> {
+        let ciphertext_start = Self::nonce_size();
+        let ciphertext_end = bytes.len() - Self::hmac_size();
+
+        Ok(Self::new(
+            bytes[..ciphertext_start].to_vec(),
+            bytes[ciphertext_start..ciphertext_end].to_vec(),
+            GenericArray::from_slice(&bytes[ciphertext_end..]),
+        ))
+    }
+
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        [&self.nonce[..], &self.ciphertext[..], &self.hmac[..]].concat()
+    }
+
+    pub(crate) fn seal<R: RngCore + CryptoRng>(
```

Comment:
Could we add a docstring on this one, especially one documenting the use of a fresh random nonce for each EnvU is necessary for the security of the enveloping scheme?